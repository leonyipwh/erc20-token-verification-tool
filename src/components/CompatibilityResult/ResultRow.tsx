import React from 'react';
import IconButton from '@mui/material/IconButton';
import Typography from '@mui/material/Typography';
import TableCell from '@mui/material/TableCell';
import TableRow from '@mui/material/TableRow';
import KeyboardArrowDownIcon from '@mui/icons-material/KeyboardArrowDown';
import KeyboardArrowUpIcon from '@mui/icons-material/KeyboardArrowUp';
import Collapse from '@mui/material/Collapse';
import { PROPERTY_STATUS } from '@netilon/differify';

const ResultRow = ({ data }: { data: any }) => {
  const [open, setOpen] = React.useState(false);
  const { inputs, name, outputs, type } = data._;

  const checkDiff = (diff: any, label: string) => {
    if (diff.changes === 0) return <></>;

    return diff._.map((item: any, index: number) => {
      // missing method
      if (item.status === PROPERTY_STATUS.DELETED) {
        return (
          <DiffExplain
            key={index}
            text={`Missing ${label} type ${item.original.type}`}
          />
        );
      }
      // unmatched type
      if (item._.type.status === PROPERTY_STATUS.MODIFIED) {
        return (
          <DiffExplain
            key={index}
            text={`Expect ${label} type ${item._.type.original}, but found ${item._.type.current}`}
          />
        );
      }
    });
  };

  return (
    <>
      <TableRow sx={{ '&:last-child td, &:last-child th': { border: 0 } }}>
        <TableCell>{type.original}</TableCell>
        <TableCell>{name.original}</TableCell>
        <TableCell align="right">
          <IconButton
            aria-label="expand row"
            size="small"
            onClick={() => setOpen(!open)}
          >
            {open ? <KeyboardArrowUpIcon /> : <KeyboardArrowDownIcon />}
          </IconButton>
        </TableCell>
      </TableRow>
      <TableRow>
        <TableCell
          style={{ paddingBottom: 0, paddingTop: 0 }}
          colSpan={4}
        >
          <Collapse
            in={open}
            timeout="auto"
            unmountOnExit
          >
            {type.status !== PROPERTY_STATUS.DELETED ? (
              <>
                {outputs && checkDiff(outputs, 'outputs')}
                {inputs && checkDiff(inputs, 'inputs')}
              </>
            ) : (
              <DiffExplain text={`Missing entire ${type.original}`} />
            )}
          </Collapse>
        </TableCell>
      </TableRow>
    </>
  );
};

const DiffExplain = ({ text }: { text: string }) => {
  return (
    <Typography
      sx={{ py: 2, px: 4 }}
      variant="subtitle2"
      gutterBottom
      component="div"
    >
      {text}
    </Typography>
  );
};

export default ResultRow;
