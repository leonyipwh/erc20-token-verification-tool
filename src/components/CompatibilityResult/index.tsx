import Table from '@mui/material/Table';
import TableBody from '@mui/material/TableBody';
import TableCell from '@mui/material/TableCell';
import TableContainer from '@mui/material/TableContainer';
import TableHead from '@mui/material/TableHead';
import TableRow from '@mui/material/TableRow';
import Paper from '@mui/material/Paper';
import Typography from '@mui/material/Typography';
import Alert from '@mui/material/Alert';
import AlertTitle from '@mui/material/AlertTitle';
import Toolbar from '@mui/material/Toolbar';
import { multiPropDiff } from '@netilon/differify';
import ResultRow from './ResultRow';

const CompatibilityResult = ({ data }: { data: multiPropDiff[] }) => {
  return (
    <>
      {!data.length ? (
        <Alert severity="success">
          <AlertTitle>Compatible</AlertTitle>
          Contract is compatible with ERC-20 token standard
        </Alert>
      ) : (
        <>
          <Alert severity="warning">
            <AlertTitle>Incompatible</AlertTitle>Contract is incompatible with ERC-20 token standard
          </Alert>
          <TableContainer
            component={Paper}
            sx={{ mt: 3 }}
          >
            <Toolbar>
              <Typography
                sx={{ flex: '1 1 100%' }}
                variant="h6"
                id="tableTitle"
                component="div"
              >
                Incompatibility
              </Typography>
            </Toolbar>
            <Table
              sx={{ minWidth: 650 }}
              aria-label="incompatibility table"
            >
              <TableHead>
                <TableRow>
                  <TableCell>Type</TableCell>
                  <TableCell>Name</TableCell>
                  <TableCell align="right" />
                </TableRow>
              </TableHead>
              <TableBody>
                {data.map((row, index) => (
                  <ResultRow
                    data={row}
                    key={index}
                  />
                ))}
              </TableBody>
            </Table>
          </TableContainer>
        </>
      )}
    </>
  );
};

export default CompatibilityResult;
