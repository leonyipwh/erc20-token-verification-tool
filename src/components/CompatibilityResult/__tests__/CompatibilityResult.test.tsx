import { render } from '@testing-library/react';
import CompatibilityResult from '../index';
import { diffMock } from '../mocks/diff.mock';

describe('<CompatibilityResult>', () => {
  it('should render', () => {
    const { asFragment } = render(<CompatibilityResult data={diffMock} />);
    expect(asFragment()).toMatchSnapshot();
  });
});
