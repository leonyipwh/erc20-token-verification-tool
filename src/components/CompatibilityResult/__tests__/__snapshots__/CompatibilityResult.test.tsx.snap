// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<CompatibilityResult> should render 1`] = `
<DocumentFragment>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation0 MuiAlert-root MuiAlert-standardWarning MuiAlert-standard css-rob5aa-MuiPaper-root-MuiAlert-root"
    role="alert"
  >
    <div
      class="MuiAlert-icon css-1ytlwq5-MuiAlert-icon"
    >
      <svg
        aria-hidden="true"
        class="MuiSvgIcon-root MuiSvgIcon-fontSizeInherit css-1vooibu-MuiSvgIcon-root"
        data-testid="ReportProblemOutlinedIcon"
        focusable="false"
        viewBox="0 0 24 24"
      >
        <path
          d="M12 5.99L19.53 19H4.47L12 5.99M12 2L1 21h22L12 2zm1 14h-2v2h2v-2zm0-6h-2v4h2v-4z"
        />
      </svg>
    </div>
    <div
      class="MuiAlert-message css-acap47-MuiAlert-message"
    >
      <div
        class="MuiTypography-root MuiTypography-body1 MuiTypography-gutterBottom MuiAlertTitle-root css-yslrua-MuiTypography-root-MuiAlertTitle-root"
      >
        Incompatible
      </div>
      Contract is incompatible with ERC-20 token standard
    </div>
  </div>
  <div
    class="MuiPaper-root MuiPaper-elevation MuiPaper-rounded MuiPaper-elevation1 MuiTableContainer-root css-1pmtgpr-MuiPaper-root-MuiTableContainer-root"
  >
    <div
      class="MuiToolbar-root MuiToolbar-gutters MuiToolbar-regular css-hyum1k-MuiToolbar-root"
    >
      <div
        class="MuiTypography-root MuiTypography-h6 css-2pbtu6-MuiTypography-root"
        id="tableTitle"
      >
        Incompatibility
      </div>
    </div>
    <table
      aria-label="incompatibility table"
      class="MuiTable-root css-quj9j6-MuiTable-root"
    >
      <thead
        class="MuiTableHead-root css-15wwp11-MuiTableHead-root"
      >
        <tr
          class="MuiTableRow-root MuiTableRow-head css-1q1u3t4-MuiTableRow-root"
        >
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Type
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-sizeMedium css-1ygcj2i-MuiTableCell-root"
            scope="col"
          >
            Name
          </th>
          <th
            class="MuiTableCell-root MuiTableCell-head MuiTableCell-alignRight MuiTableCell-sizeMedium css-1azl6jz-MuiTableCell-root"
            scope="col"
          />
        </tr>
      </thead>
      <tbody
        class="MuiTableBody-root css-apqrd9-MuiTableBody-root"
      >
        <tr
          class="MuiTableRow-root css-34nofg-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            function
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            transfer
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            colspan="4"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
        <tr
          class="MuiTableRow-root css-34nofg-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            event
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
          >
            Approval
          </td>
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-alignRight MuiTableCell-sizeMedium css-177gid-MuiTableCell-root"
          >
            <button
              aria-label="expand row"
              class="MuiButtonBase-root MuiIconButton-root MuiIconButton-sizeSmall css-1pe4mpk-MuiButtonBase-root-MuiIconButton-root"
              tabindex="0"
              type="button"
            >
              <svg
                aria-hidden="true"
                class="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
                data-testid="KeyboardArrowDownIcon"
                focusable="false"
                viewBox="0 0 24 24"
              >
                <path
                  d="M7.41 8.59 12 13.17l4.59-4.58L18 10l-6 6-6-6 1.41-1.41z"
                />
              </svg>
              <span
                class="MuiTouchRipple-root css-8je8zh-MuiTouchRipple-root"
              />
            </button>
          </td>
        </tr>
        <tr
          class="MuiTableRow-root css-1q1u3t4-MuiTableRow-root"
        >
          <td
            class="MuiTableCell-root MuiTableCell-body MuiTableCell-sizeMedium css-1ex1afd-MuiTableCell-root"
            colspan="4"
            style="padding-bottom: 0px; padding-top: 0px;"
          />
        </tr>
      </tbody>
    </table>
  </div>
</DocumentFragment>
`;
